auth/registry.go

```diff
@@ -0,0 +1,210 @@
+	if strings.Count(address, ":") > 0 {
+		tmp := address[strings.LastIndex(address, ":"):]
+
+		// If it's a tag, it may contains a "."
+		if strings.Count(tmp, ".") > 0 {
+			address = address[:strings.LastIndex(address, ":")]
+		}
+	}
+
+	addressPart := strings.SplitN(address, "/", 2)
+	domain := addressPart[0]
+
+	switch {
+	// Local registry
+	// E.g., localhost:5000
+	case strings.Contains(domain, "localhost"):
+		return domain, nil
+	// If the address is only an image name without any "."
+	// E.g., ubuntu, alpine, redis
+	case len(address) == 1 && !strings.Contains(domain, "."):
+		return defaultDockerDomain, nil
+	// If the address contains an image without "." nor ":"
+	// E.g., bitnami/redis
+	case !strings.Contains(address, ".") && !strings.Contains(address, ":"):
+		return defaultDockerDomain, nil
+	// If the address is docker hub related.
+	// E.g., registry-1.docker.io, index.docker.io
+	case strings.Contains(address, "docker.io"):
+		return defaultDockerDomain, nil
+	// Private registry or other well formatted address.
+	case strings.Contains(domain, "."):
+		return domain, nil
+	default:
+		return "", fmt.Errorf("failed parsing [%s], expected address format: [%s]", domain, "registrydomain.extension")
+	}
+}
+
+// RemoveCredential suppress the credential bind to the given address.
+// Nothing happens if the address has no credential.
+func (r *RegistryAuthProvider) RemoveCredential(address string) error {
+	address, err := parseAuthAddress(address)
+	if err != nil {
+		return err
+	}
+
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	delete(r.credentials, address)
+	return nil
+}
+
+func (r *RegistryAuthProvider) Register(server *grpc.Server) {
+	bkauth.RegisterAuthServer(server, r)
+}
+
+func (r *RegistryAuthProvider) credential(domain string) *bkauth.CredentialsResponse {
+	// Update default DNS of Docker Hub registry to short name.
+	if domain == "registry-1.docker.io" || domain == "index.docker.io" {
+		domain = defaultDockerDomain
+	}
+
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	for authAddress, credential := range r.credentials {
+		if authAddress == domain {
+			return credential
+		}
+	}
+
+	return nil
+}
+
+// Credentials retrieves credentials of the requested address.
+// It searches in the memory map for the standardize address.
+//
+// If the address isn't registered in the memory map, it will search
+// on DockerAuthProvider.
+func (r *RegistryAuthProvider) Credentials(ctx context.Context, req *bkauth.CredentialsRequest) (*bkauth.CredentialsResponse, error) {
+	memoryCredential := r.credential(req.GetHost())
+	if memoryCredential != nil {
+		return memoryCredential, nil
+	}
+
+	return r.dockerAuthProvider.Credentials(ctx, req)
+}
+
+func (r *RegistryAuthProvider) FetchToken(ctx context.Context, req *bkauth.FetchTokenRequest) (*bkauth.FetchTokenResponse, error) {
+	memoryCredential := r.credential(req.GetHost())
+	if memoryCredential != nil {
+		return nil, status.Errorf(codes.Unavailable, "secret is store in memory")
```

Comment:
I would have thought it'd be okay to just fallback to trying the dockerAuthProvider in any case where the in memory provider isn't valid. What's the reason for returning an error here?

Reply:
Actually, this is what is happening, if the credential exists in memory, it will return an error to skip the auth provider test.
If I do not try that, it's considered as if `dockerAuthProvider` failed.

It's a hacky way to manage the situation but that is the only one that could both manage `DockerAuth` and our custom secret management.